src/Grids/grid_generation.jl

```diff
@@ -113,6 +113,10 @@ function generate_coordinate(FT, topo::AT, N, H, node_interval::Tuple{<:Number,
     F = range(FT(F₋), FT(F₊), length = TF)
     C = range(FT(C₋), FT(C₊), length = TC)
 
+    # See: https://github.com/CliMA/Oceananigans.jl/issues/3870#issuecomment-2445532635
+    F = StepRangeLen{FT, FT, FT, Int}(F)
+    C = StepRangeLen{FT, FT, FT, Int}(C)
```

Comment:
I don't think it is correct to describe this change as "making the grid coordinate ranges type-safe".

Perhaps more conservatively, we only require the _first_ type parameter of `StepRangeLen` to be `FT`. The second two should be twice precision.

Now, if we want to support non-standard ranges, we can perhaps consider that. But I suspect there is something going on that isn't completely understood here.

Reply:
This is discussed on #3870, but just to express the problem here: we _expect_ the ranges to have the type `StepRangeLen{Float32, Float64, Float64, Int}` for a `Float32` grid. Therefore, this PR does some damage by producing an unexpected range type.

It is unexpected or a bug that a range of type `StepRangeLen{Float32, Float64, Float64, Int}` would produce `Float64`. Therefore the first course of action is to _verify_ that `StepRangeLen{Float32, Float64, Float64, Int}` is producing `Float64` --- on either CPU or GPU.

I have a hunch that `StepRangeLen` is somehow adapted for GPU incorrectly. Or, if it is intentional, then I think we should fix the output of `xnode`, `ynode`, `znode`...

If there's good motivation for reducing the precision of ranges (either in addition or alternatively to the above suggestion) then I think we can entertain it. I'm not sure we want to hardcode this change though, it might be better to provide it as an option. The imprecision of ranges is already a bit annoying.

Reply:
Since this makes the test pass and the MWE in #3870 not error, maybe the hint is that `StepRangeLen` just needs to be adapted for the GPU?

But yeah I'm not sure why `StepRangeLen{Float32, Float64, Float64, Int}` seems to produce `Float64` numbers in GPU kernels. Perhaps this is worth opening an issue on CUDA.jl with a MWE?

Reply:
Yes, though it is not for us to adapt (that would be type piracy) because we own neither `Adapt` nor `StepRangeLen`. Plus it seems to be adapted, so what is happening...

https://github.com/JuliaGPU/Adapt.jl/blob/5ef7c5329609df7ffb5b19942d6747b3dcc162c2/src/base.jl#L79-L80